### PR TITLE
Audio PR7 - Audio augmentations 2 (MFBB and Normalize)

### PR DESCRIPTION
**Pull request by SundarRajan28**

### CODE REVIEW

File: rocAL/include/api/rocal_api_augmentation.h [hunk 1]
```diff
@@ -1277,4 +1277,51 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
...
 
+/*! \brief Performs mean-stddev normalization on images.
+ * \ingroup group_rocal_augmentations
+ * \param [in] context Rocal context
+ * \param [in] input Input Rocal tensor
+ * \param [in] axes axes list for tensor normalization
+ * \param [in] mean mean value (specified for each channel) for tensor normalization
+ * \param [in] std_dev standard deviation value (specified for each channel) for tensor normalization
+ * \param [in] is_output is the output tensor part of the graph output
+ * \param [in] scale scale value (specified for each channel) for tensor normalization
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h:1288]: Change the description, do we need to have for each channel

Comment (fiona-gladwin) [reply rocAL/include/api/rocal_api_augmentation.h:1288]: Same for scale shift ?

File: rocAL/include/api/rocal_api_augmentation.h [hunk 2]
```diff
@@ -1277,4 +1277,51 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
...
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalNormalize(RocalContext context, RocalTensor input,
+                                                     std::vector<unsigned> &axes,
+                                                     std::vector<float> &mean,
+                                                     std::vector<float> &std_dev,
+                                                     bool is_output,
+                                                     float scale = 1.0, float shift = 0.0,
+                                                     RocalTensorOutputType output_datatype = ROCAL_FP32);
+
+/*! \brief Applies mel-filter bank augmentation on the given input tensor
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_augmentation.h:1301]: This is for spectrogram outputs only right? Should we mention it?

File: rocAL/include/api/rocal_api_types.h [hunk 3]
```diff
@@ -424,4 +424,9 @@ enum RocalOutOfBoundsPolicy {
     ROCAL_ERROR
 };
 
+enum RocalMelScaleFormula {
+    SLANEY = 0,  // Follows Slaney’s MATLAB Auditory Modelling Work behavior
```
Comment (fiona-gladwin) [inline rocAL/include/api/rocal_api_types.h]: Please add ROCAL_ suffix to enums

Comment (fiona-gladwin) [reply rocAL/include/api/rocal_api_types.h]: Also add brief / description  refer to the previous enums

File: rocAL/include/augmentations/effects_augmentations/node_normalize.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    void update_node() override{};
+
+   private:
+    int _axis_mask = 0;
+    uint _compute_mean, _compute_stddev;
+    vx_array _mean_vx_array, _stddev_vx_array;
+    std::vector<unsigned> _axes;
+    std::vector<float> _mean, _std_dev;
+    float _scale, _shift;
+    std::vector<std::vector<uint32_t>> _normalize_roi;
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/effects_augmentations/node_normalize.h]: Unused variable

File: rocAL/include/augmentations/effects_augmentations/node_normalize.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+   protected:
+    void create_node() override;
+    void update_node() override{};
+
+   private:
+    int _axis_mask = 0;
+    uint _compute_mean, _compute_stddev;
+    vx_array _mean_vx_array, _stddev_vx_array;
+    std::vector<unsigned> _axes;
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/effects_augmentations/node_normalize.h]: Unused variable

File: rocAL/include/augmentations/effects_augmentations/node_normalize.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    void init(std::vector<unsigned> &axes, std::vector<float> &mean, std::vector<float> &std_dev, float scale, float shift);
+
+   protected:
+    void create_node() override;
+    void update_node() override{};
+
+   private:
+    int _axis_mask = 0;
+    uint _compute_mean, _compute_stddev;
+    vx_array _mean_vx_array, _stddev_vx_array;
```
Comment (fiona-gladwin) [inline rocAL/include/augmentations/effects_augmentations/node_normalize.h]:
```suggestion
    uint _compute_mean, _compute_stddev;
    vx_array _mean_vx_array, _stddev_vx_array;
```

All these 4 variables can be made local

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        mean_stddev_array_size = std::max(mean_stddev_array_size, totalElements);
+    }
+    std::vector<float> mean_vec, stddev_vec;
+    mean_vec.resize(_batch_size * mean_stddev_array_size, 0);
+    stddev_vec.resize(_batch_size * mean_stddev_array_size, 0);
+
+    if (!_compute_mean && !_compute_stddev) {
+        for (uint i = 0; i < _batch_size; i++) {
+            for (int j = 0; j < mean_stddev_array_size; j += _mean.size()) {
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:61]: Can we add a check if mean and std dev is of the same size

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    status |= vxAddArrayItems(_mean_vx_array, mean_vec.size(), mean_vec.data(), sizeof(vx_float32));
+    if (status != 0)
+        THROW(" vxAddArrayItems failed in the normalize node (vxExtRppNormalize)  node: " + TOSTR(status) + "  " + TOSTR(status))
+
+    _stddev_vx_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, stddev_vec.size());
+    status |= vxAddArrayItems(_stddev_vx_array, stddev_vec.size(), stddev_vec.data(), sizeof(vx_float32));
+    if (status != 0)
+        THROW(" vxAddArrayItems failed in the normalize node (vxExtRppNormalize)  node: " + TOSTR(status) + "  " + TOSTR(status))
+
+    vx_scalar axis_mask = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &_axis_mask);
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp]: Add _vx suffix to all the vx_scalar variables

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    if (_node)
+        return;
+
+    _compute_mean = _mean.size() ? 0 : 1;
+    _compute_stddev = _std_dev.size() ? 0 : 1;
+
+    int mean_stddev_array_size = 1;
+    auto nDim = _inputs[0]->info().num_of_dims() - 1;
+    uint axis[nDim];
+    for (unsigned i = 0; i < _batch_size; i++) {
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:46]: Please add comment explaining what this snippet does

File: rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "exception.h"
+
+MelFilterBankNode::MelFilterBankNode(const std::vector<Tensor *> &inputs, const std::vector<Tensor *> &outputs) : Node(inputs, outputs) {}
+
+void MelFilterBankNode::create_node() {
+    if (_node)
+        return;
+
+    vx_scalar freq_high = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_FLOAT32, &_freq_high);
```
Comment (fiona-gladwin) [inline rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp]: Add _vx suffix to vx_acalar variables

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 11]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+    if ((p_context == nullptr) || (p_input == nullptr)) {
+        ERR("Invalid ROCAL context or invalid input tensor")
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_datatype = static_cast<RocalTensorDataType>(output_datatype);
+        TensorInfo output_info = input->info();
+        output_info.set_data_type(op_tensor_datatype);
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2524]: Do we support all data types

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 12]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
+        TensorInfo output_info = input->info();
+        std::vector<size_t> max_dims = output_info.max_shape();
+        int max_frame = max_dims[1];
+        max_frame = std::max(0, max_frame);
+        std::vector<size_t> dims = output_info.dims();
+        dims[1] = nfilter;
+        dims[2] = max_frame;
+        output_info.set_dims(dims);
+        output_info.set_data_type(op_tensor_data_type);
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2566]: Do we support different data types?

Comment (fiona-gladwin) [reply rocAL/source/api/rocal_api_augmentation.cpp:2566]: If not add a comment

File: rocAL/source/api/rocal_api_augmentation.cpp [hunk 13]
```diff
@@ -2504,3 +2504,68 @@ rocalSlice(
...
+    RocalTensorOutputType output_datatype) {
+    Tensor* output = nullptr;
+    if ((p_context == nullptr) || (p_input == nullptr)) {
+        ERR("Invalid ROCAL context or invalid input tensor")
+        return output;
+    }
+    auto context = static_cast<Context*>(p_context);
+    auto input = static_cast<Tensor*>(p_input);
+    try {
+        RocalTensorDataType op_tensor_data_type = (RocalTensorDataType)output_datatype;
```
Comment (fiona-gladwin) [inline rocAL/source/api/rocal_api_augmentation.cpp:2555]: Use static cast

File: rocAL_pybind/examples/rocAL_api_audio_unittest.py [hunk 14]
```diff
@@ -76,9 +78,9 @@ def verify_output(output_list, rocal_data_path, roi_tensor, test_results, case_n
     roi_data = roi_tensor.detach().numpy()
     buffer_size = roi_data[0] * roi_data[1]
     matched_indices = 0
-    for i in range(roi_data[1]):
-        for j in range(roi_data[0]):
-            ref_val = data_array[i * roi_data[0] + j]
+    for i in range(roi_data[0]):
```
Comment (fiona-gladwin) [inline rocAL_pybind/examples/rocAL_api_audio_unittest.py]: This should be added in Spectrogram PR itself

File: rocAL_pybind/examples/rocAL_api_audio_unittest.py [hunk 15]
```diff
@@ -300,16 +347,20 @@ def main():
                                                     path=downmix_audio_path if qa_mode else audio_path, file_list="" if qa_mode else file_list, downmix=True)
         if case_name == "to_decibels":
             audio_pipeline = to_decibels_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
```
Comment (fiona-gladwin) [inline rocAL_pybind/examples/rocAL_api_audio_unittest.py]: Why are you removing this in this PR>?

File: tests/cpp_api_tests/rocAL_audio_unittests/rocAL_audio_unittests.cpp [hunk 16]
```diff
@@ -109,16 +109,16 @@ bool verify_output(float *dst_ptr, long int frames, std::string case_name, int m
 
     if (fin.fail()) {
         std::cout << "Error: Failed to read from the input binary file\n";
-        return 1;
+        return 0;
```
Comment (fiona-gladwin) [inline tests/cpp_api_tests/rocAL_audio_unittests/rocAL_audio_unittests.cpp]: Why is this changed?

File: tests/cpp_api_tests/rocAL_audio_unittests/rocAL_audio_unittests.cpp [hunk 17]
```diff
@@ -109,16 +109,16 @@ bool verify_output(float *dst_ptr, long int frames, std::string case_name, int m
...
     }
 
     fin.close();
 
     int matched_indices = 0;
-    for (int i = 0; i < max_samples; i++) {
-        for (int j = 0; j < frames; j++) {
+    for (int i = 0; i < frames; i++) {
```
Comment (fiona-gladwin) [inline tests/cpp_api_tests/rocAL_audio_unittests/rocAL_audio_unittests.cpp]: All these changes should be done in Spectrogram PR itself, Please sync with Swetha and add these in Spec PR

File: rocAL/include/api/rocal_api_augmentation.h [hunk 18]
```diff
@@ -1277,4 +1277,51 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
                                                  RocalOutOfBoundsPolicy policy = RocalOutOfBoundsPolicy::ROCAL_ERROR,
                                                  RocalTensorOutputType output_datatype = ROCAL_FP32);
 
+/*! \brief Performs mean-stddev normalization on images.
```
Comment (sampath1117) [inline rocAL/include/api/rocal_api_augmentation.h:1280]: change the description as 
`Performs mean-stddev normalization on input tensors
`

File: rocAL/include/api/rocal_api_augmentation.h [hunk 19]
```diff
@@ -1277,4 +1277,51 @@ extern "C" RocalTensor ROCAL_API_CALL rocalSlice(RocalContext context,
...
+ * \ingroup group_rocal_augmentations
+ * \param [in] context Rocal context
+ * \param [in] input Input Rocal tensor
+ * \param [in] axes axes list for tensor normalization
+ * \param [in] mean mean value (specified for each channel) for tensor normalization
```
Comment (sampath1117) [inline rocAL/include/api/rocal_api_augmentation.h:1285]: since it is a generic normalize, better to not to mention channels here

File: rocAL/include/api/rocal_api_augmentation.h (hunk 2 above)
Comment (sampath1117) [reply rocAL/include/api/rocal_api_augmentation.h:1301]: Yes better to mention spectrogram as input here

File: rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (sampath1117) [inline rocAL/source/augmentations/audio_augmentations/node_mel_filter_bank.cpp]: Please change copyright

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        int totalElements = 1;
+        unsigned *tensor_shape = _inputs[0]->info().roi()[i].end;
```
Comment (sampath1117) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:48]: we are mixing camel case and snake case here
Please check and change

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+        for (uint j = 0; j < nDim; j++) {
+            axis[j] = ((_axis_mask & (int)(pow(2, j))) >= 1) ? 1 : 0;
+            totalElements *= !axis[j] ? tensor_shape[j] : 1;
+        }
+        mean_stddev_array_size = std::max(mean_stddev_array_size, totalElements);
+    }
+    std::vector<float> mean_vec, stddev_vec;
+    mean_vec.resize(_batch_size * mean_stddev_array_size, 0);
+    stddev_vec.resize(_batch_size * mean_stddev_array_size, 0);
```
Comment (sampath1117) [inline rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:57]: In new kernel, if mean and stddev passed from user
we need not replicate the same samples for all samples in batch and that allocation also is not needed

Like if mean is passed as [100, 110, 120]
This can be passed as is to RPP. Need replicate the same values for all samples in batch

File: rocAL/source/augmentations/effects_augmentations/node_normalize.cpp (hunk 7 above)
Comment (sampath1117) [reply rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:61]: Yes
Better to add a validation check verifying if mean and std passed is of same size

File: rocAL_pybind/examples/rocAL_api_audio_unittest.py [hunk 23]
```diff
@@ -300,16 +347,20 @@ def main():
...
             audio_pipeline = tensor_add_tensor_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
         if case_name == "tensor_mul_scalar":
             audio_pipeline = tensor_mul_scalar_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
+        if case_name == "slice":
+            audio_pipeline = non_silent_region_and_slice(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
+        if case_name == "non_silent_region":
+            audio_pipeline = non_silent_region(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
+        if case_name == "mel_filter_bank":
+            audio_pipeline = mel_filter_bank_pipeline(batch_size=batch_size, num_threads=num_threads, device_id=device_id, rocal_cpu=rocal_cpu, path=audio_path, file_list=file_list)
+        if case_name == "normalize":
```
Comment (sampath1117) [inline rocAL_pybind/examples/rocAL_api_audio_unittest.py]: change all conditions to use else if, instead of using if for all

**Review by sampath1117 — COMMENTED** [review]
Please address review comments
